### PR TITLE
regl example demo

### PR DESCRIPTION
**Pull request by kevzettler**
Opening this for discussion until its ready and will move to regl-examples

### CODE REVIEW

**Comment by kevzettler** [comment]
@chinedufn I'm currently blocked on this and hope you have a minute to take a look.

seeing an error: 
```
Error: (regl) bad data or missing for uniform "boneTransQuaternions[0]".  invalid vector, should have length 4 in command http://localhost:9966/regl.js:172:23 called from http://localhost:9966/regl.js:327:3
```

pretty sure that is stemming from: 
https://github.com/chinedufn/skeletal-animation-system/pull/3/files#diff-0d1c6153ea57c6bded9821fb0f8e79e1R144

That uniform gets passed into the regl shader via:
1. https://github.com/chinedufn/skeletal-animation-system/pull/3/files#diff-0d1c6153ea57c6bded9821fb0f8e79e1R291

2. https://github.com/chinedufn/skeletal-animation-system/pull/3/files#diff-0d1c6153ea57c6bded9821fb0f8e79e1R254

I'm not sure what the proper way to pass in `boneTransQuaternions` to the shader is.

Looking at the old code:
`gl.uniform4fv(bufferData.shader.uniforms['boneTransQuaternions' + jointNum].location, drawOpts.transQuaternions[jointNum])`

I'm not sure how to translate that to regl style.

**Comment by chinedufn** [comment]
@kevzettler sadly I'm not familiar, so tagging @erkaman and @mikolalysenko in case any of them have a quick answer about how to format arrays

**Comment by kevzettler** [comment]
pretty sure the problem is: https://github.com/regl-project/regl/issues/373#issuecomment-258020073

**Comment by kevzettler** [comment]
@chinedufn I made some updates here. The code is not throwing any errors. Unfortunately, it is rendering an empty canvas. I could use your review on the shader code. I believe I have all the data flowing correctly through regl now to the shader code but think some values off. I'm suspecting I may have broke the camera, lighting, or texture assumptions.

File: demo/regl.js [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+  frag: `
+    precision mediump float;
+    varying vec3 vLightWeighting;
+    void main (void) {
+      gl_FragColor = vec4(vLightWeighting, 1.0);
+    }`,
+
+  attributes: {
+    position: simplicalCowboy.positions,
+    aVertexPosition: cowboy.vertexPositions,
```
Comment (chinedufn) [inline demo/regl.js]: @kevzettler from what I'm seeing in their docs I think `regl` requires you to wrap these attributes in `regl.buffer()`

image

**Comment by kevzettler** [comment]
Ok @chinedufn Thanks for pointing that out. I've updated the attributes to use `regl.buffer` and trying to match the webGL methods from `draw-model` for example:

`draw-model` WebGL:
```javascript
    bufferData.aVertexNormal = createBuffer(gl, 'ARRAY_BUFFER', Float32Array, vertexData.vertexNormals)
    gl.bindBuffer(gl.ARRAY_BUFFER, bufferData.aVertexNormal)
    gl.enableVertexAttribArray(bufferData.shader.attributes.aVertexNormal.location)
    gl.vertexAttribPointer(bufferData.shader.attributes.aVertexNormal.location, 3, gl.FLOAT, false, 0, 0)
```

Regl buffers:
```javascript
  attributes: {
    aVertexNormal: {
      buffer: regl.buffer({
        data: vertexData.vertexNormals,
        type: 'float32'
      }),
      size: 3
    },
}
```


This setup is now throwing some WebGL errors:
```
[.Offscreen-For-WebGL-0x7f8c5e131000]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 1
```

Assuming `attribute 1` is referring to `aVertexNormal` I can change the `size` of `aVertexNormal` to 4 and it will then throw :

```
[.Offscreen-For-WebGL-0x7f8c5e131000]GL ERROR :GL_INVALID_OPERATION : glDrawElements: attempt to access out of range vertices in attribute 0
```

Which I can't seem to remedy. Hope you have another insight I'm overlooking appreciate all the help

File: demo/regl.js [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+      buffer: regl.buffer({
+        data: vertexData.vertexJointAffectors,
+        type: 'float32'
+      }),
+      size: 4
+    },
+    
+    aJointWeight: {
+      buffer: regl.buffer({
+        buffer: vertexData.vertexJointWeights,
```
Comment (chinedufn) [inline demo/regl.js]: this says `buffer` but I'm thinking you meant `data`?

Comment (kevzettler) [inline demo/regl.js]: Good catch. Pushed a fix.

**Comment by kevzettler** [comment]
Hey @chinedufn, update on this. I made some code changes to try and 1 to 1 match the data in the existing demo that gets passed to drawModel. Like adding the texture in and camera perspective. Still getting a blank screen rendering with no error output. I suspect it has to do with the Uniform arrays: `boneRotQuaternions`, and `boneTransQuaternions` It seemss regl might be sensitive to this Uniform array pattern:

I filed an issue for expanding on what I'm seeing:
https://github.com/regl-project/regl/issues/404

I'm hoping you could give this another eyeball to make sure I'm not missing anything. Also wondering if you have any thoughts on tweaking the shader at all to pass the values in as attribute. Appreciate the help on this. I might have to shelve this for the moment to work on other commitments

File: demo/regl.js [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+      elements: regl.elements({
+        data: vertexData.vertexPositionIndices,
+        type: 'uint16',
+        primitive: 'triangles'
+      }),
+      
+      uniforms: Object.assign({
+        boneRotQuaternions: regl.prop('boneRotQuaternions'),
+        boneTransQuaternions: regl.prop('boneTransQuaternions'),
+        uMVMatrix: drawOpts.perspective,
```
Comment (chinedufn) [inline demo/regl.js]: Looks like you might have your model view matrix and perspective matrix reversed

**Comment by chinedufn** [comment]
@kevzettler I took a look at your branch locally. Looks like if you switch your model-view-matrix and your perspective matrix your bones will start to work. The issue was that they were working, but you couldn't see them because your camera was accidentally pointing somewhere else.

image

Looks like the textures are a bit off, but it's animating!

Awesome work on all of this!

---

With regards to potentially needing to shelve this -> no biggie, life happens indeed

**Comment by kevzettler** [comment]
@chinedufn I just made and pushed the same updates after seeing your last comment! Very cool! Thanks for the peer review. The animation is really jittery for me like the delta-time argument is off or something. Any thoughts on that?. I'll look some more at the texture probably another slight mistake there.

File: demo/regl.js [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        return accum;
+      }, {}), new Uint32Array(18).reduce((accum, value, index) => {
+        accum['boneTransQuaternions['+index+']'] = regl.prop('boneTransQuaternions['+index+']');
+        return accum;
+      }, {})),
+    });
+
+    var currentTime = 0;
+    regl.frame(({time}) => {
+      currentTime += time;
```
Comment (chinedufn) [inline demo/regl.js]: Looks like it's moving through the keyframes so quickly that it appears to jitter.

In the original demo we divide the time delta by `1000`

ex: https://github.com/chinedufn/skeletal-animation-system/blob/master/demo/demo.js#L109

So hopefully something like that can help with the jitter

**Comment by kevzettler** [comment]
@chinedufn I fixed up the texture and cleaned up the code a bit and matched Standard js style. The demo is not hooked up to UI triggers like the standard demo. Let me know if you have any thoughts on further improving the code. Feel free to merge this and iterate on, or leave open.

**Comment by chinedufn** [comment]
@kevzettler again, amazing work on this. Excited for your work to start helping other people learn more about implementing skeletal animation.

I've added you as an owner of `skeletal-animation-system`. Thanks for all of your help! Feel free to merge this in.

**Pull request merged**